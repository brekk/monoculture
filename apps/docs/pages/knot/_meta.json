{
  "packages-knot-knot.js": "knot"
}