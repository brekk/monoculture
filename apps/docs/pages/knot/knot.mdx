# knot



## nthIndex

Slice a string by counted delimiters

### Usage
```js test=true
expect(
  nthIndex('/', -5, 'a/b/c/d/e/f/g/h/i/j')
).toEqual("f/g/h/i/j")
expect(
  nthIndex('/', 5, 'a/b/c/d/e/f/g/h/i/j')
).toEqual("a/b/c/d/e")
```

## strepeat

A simple memoized utility for repeating a string and joining the array.

### Usage
```js test=true
expect(strepeat('=', 5)).toEqual('=====')
expect(strepeat('/', -1)).toEqual('')
```