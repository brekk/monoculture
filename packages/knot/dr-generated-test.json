[
  {
    "slugName": "knot",
    "pageSummary": "",
    "filename": "knot.js",
    "comments": [
      {
        "start": 53,
        "end": 62,
        "summary": "A simple memoized utility for repeating a string and joining the array.",
        "links": [],
        "fileGroup": "",
        "addTo": "",
        "structure": {
          "name": "strepeat",
          "example": "```js test=true\nimport { strepeat } from '../knot'\n// drgen-import-above\nexpect(strepeat('=', 5)).toEqual('=====')\n```"
        },
        "keywords": [
          "@example",
          "@name"
        ],
        "filename": "knot.js"
      }
    ],
    "order": 0,
    "links": [],
    "testPath": "knot.spec.js"
  }
]