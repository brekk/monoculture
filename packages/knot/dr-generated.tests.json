[
  {
    "slugName": "knot",
    "pageSummary": "",
    "filename": "knot.js",
    "comments": [
      {
        "start": 48,
        "end": 57,
        "summary": "Slice a string by counted delimiters",
        "links": [],
        "fileGroup": "",
        "addTo": "",
        "structure": {
          "name": "nthIndex",
          "example": "```js test=true\nexpect(\n  nthIndex('/', -5, 'a/b/c/d/e/f/g/h/i/j')\n).toEqual(\"f/g/h/i/j\")\n```"
        },
        "keywords": [
          "@example",
          "@name"
        ],
        "filename": "knot.js"
      },
      {
        "start": 62,
        "end": 69,
        "summary": "A simple memoized utility for repeating a string and joining the array.",
        "links": [],
        "fileGroup": "",
        "addTo": "",
        "structure": {
          "name": "strepeat",
          "example": "```js test=true\nexpect(strepeat('=', 5)).toEqual('=====')\n```"
        },
        "keywords": [
          "@example",
          "@name"
        ],
        "filename": "knot.js"
      }
    ],
    "order": 0,
    "links": [],
    "testPath": "knot.spec.js"
  }
]