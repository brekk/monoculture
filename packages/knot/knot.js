import {
  toLower,
  curry,
  replace,
  join,
  pipe,
  range,
  reduce,
  split,
} from 'ramda'
import { NEWLINE, SPACE, EMPTY, TAB } from './constants'
export * from './constants'

const makeSplitJoinPair = z => [split(z), join(z)]

export const [words, unwords] = makeSplitJoinPair(SPACE)
export const [lines, unlines] = makeSplitJoinPair(NEWLINE)
export const [chars, unchars] = makeSplitJoinPair(EMPTY)
export const [tabs, untabs] = makeSplitJoinPair(TAB)
export const prepend = curry((pre, x) => `${pre}${x}`)
export const append = curry((post, x) => `${x}${post}`)

export const cake = curry(
  (bottom, top, layer, x) => pipe(join(layer), prepend(top), append(bottom))(x)
  // `${top}${join(layer, x)}${bottom}`
)
const MARKDOWN_LIST_ITEM = ' - '
export const markdownTabs = cake(
  NEWLINE,
  `\n${MARKDOWN_LIST_ITEM}`,
  MARKDOWN_LIST_ITEM
)

export const nthIndexOf = curry(function _nthIndexOf(delim, n, input) {
  return pipe(
    range(0),
    reduce((x, _) => input.indexOf(delim, x + 1), -1),
    z => input.slice(0, z)
  )(n)
})

export const nthLastIndexOf = curry(function _nthLastIndexOf(delim, n, input) {
  return pipe(
    range(0),
    reduce((x, _) => input.lastIndexOf(delim, x - 1), input.length),
    z => input.slice(z + 1)
  )(Math.abs(n))
})
/**
 * Slice a string by counted delimiters
 * @name nthIndex
 * @example
 * ```js test=true
 * expect(
 *   nthIndex('/', -5, 'a/b/c/d/e/f/g/h/i/j')
 * ).toEqual("f/g/h/i/j")
 * expect(
 *   nthIndex('/', 5, 'a/b/c/d/e/f/g/h/i/j')
 * ).toEqual("a/b/c/d/e")
 * ```
 */
export const nthIndex = curry(function _nthIndex(delim, n, input) {
  return (n > 0 ? nthIndexOf : nthLastIndexOf)(delim, n, input)
})

/**
 * A simple memoized utility for repeating a string and joining the array.
 * @name strepeat
 * @example
 * ```js test=true
 * expect(strepeat('=', 5)).toEqual('=====')
 * expect(strepeat('/', -1)).toEqual('')
 * ```
 */
export const strepeat = curry(function _strepeat(toRepeat, x) {
  return x > 0 ? toRepeat.repeat(x) : ''
})

/**
 * Capitalize a string
 * @name capitalize
 * @example
 * ```js test=true
 * expect(capitalize("nice")).toEqual("Nice")
 * expect(capitalize("")).toEqual("")
 * ```
 */
export const capitalize = raw =>
  raw.length ? `${raw[0].toUpperCase()}${raw.slice(1)}` : ''

/**
 * Take PascalCase and kebabCase inputs and replace them with slug-case
 * @name slugWord
 * @example
 * ```js test=true
 * expect(slugWord('CoolFuckingShit')).toEqual('cool-fucking-shit')
 * expect(slugWord('hoorayNiceLife')).toEqual('hooray-nice-life')
 * expect(slugWord('Do nothingCool ever')).toEqual('do nothing-cool ever')
 * expect(
 *   slugWord('src/components/homepage/AugmentedDetailsDumbComponent')
 * ).toEqual(
 *   'src-components-homepage-augmented-details-dumb-component'
 * )
 * ```
 */
export const slugWord = pipe(
  replace(/[A-Z]/g, match => `-` + match),
  replace(/\//g, '-'),
  replace(/--/g, '-'),
  replace(/^-/, ''),
  toLower
)
