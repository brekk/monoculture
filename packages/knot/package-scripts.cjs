const sd = (script, description = '') =>
  !!description ? { script, description } : { script }
module.exports = {
  scripts: {
    autogen: {
      ...sd(
        'nps -c ./package-scripts.cjs autogen.tests',
        'automatically regenerate stuff!'
      ),
      tests: sd(
        `drgen -i ./package.json -o .`,
        'automatically generate tests!'
      ),
    },
    clean: sd('rm -r dist', 'clean the build'),
    lint: sd('eslint --fix .', 'lint!'),
    test: {
      ...sd('jest', 'test!'),
      watch: sd('jest --watch', 'test with watch-mode!'),
    },
  },
}
