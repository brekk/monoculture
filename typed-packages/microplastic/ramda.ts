// so remeda doesn't seem like it actually solves much at first blush

import { curry } from 'ramda'

interface Transformer<F, T> {
  (x: F): T
}

interface EqualishBy<F, T> {
  (t: Transformer<F, T>, expected: T, x: F): boolean
  (t: Transformer<F, T>, expected: T): (x: F) => boolean
  (t: Transformer<F, T>): (expected: T, x: F) => boolean
}

export function equalishBy<F, T>(a: Transformer<F, T>, b: T, c: F) {
  const raw: EqualishBy<F, T> = curry((transform, expected, x) => {
    const t = transform(x)
    return t === expected
  })
  return raw(a, b, c)
}
