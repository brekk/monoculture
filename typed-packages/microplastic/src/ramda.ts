import { curry } from 'ramda'

interface Transformer<F, T> {
  (arg: F): T
}

export interface EqualishBy<F, T> {
  (transform: Transformer<F, T>, expected: T, x: F): boolean
  (transform: Transformer<F, T>, expected: T): (x: F) => boolean
  (transform: Transformer<F, T>): (expected: T, x: F) => boolean
}

export const equalishBy = curry(
  <F, T>(transform: Transformer<F, T>, expected: T, x: F) =>
    transform(x) === expected
)
