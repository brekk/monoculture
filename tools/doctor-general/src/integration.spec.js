// import { cwd } from 'node:process'
import { pipe, map, join, identity as I, propOr, both } from 'ramda'
import {
  join as pathJoin,
  relative,
  resolve as pathResolve,
  sep as SEPARATOR,
} from 'node:path'
import { fork } from 'fluture'
import stripAnsi from 'strip-ansi'
import {
  readFile,
  flexecaWithOptionsAndCancel,
  removeFile,
  rimraf,
} from 'file-system'
import { nthIndex } from 'knot'

test('smoke', () => {
  expect(true).toBeTruthy()
})

const run = pipe(
  flexecaWithOptionsAndCancel(() => {}, '../dist/doctor-general.cjs', {
    cwd: __dirname,
  }),
  map(propOr('', 'stdout'))
)

const goodrun = (args, expectation = I) =>
  test(`doctor-general ${join(' ')(args)}`, done => {
    pipe(
      run,
      fork(done)(z => {
        expectation(done, z)
      })
    )(args)
  })

const inFix = x =>
  nthIndex(SEPARATOR, -2, pathResolve(__dirname, '../fixture/' + x))

const GENERATED = inFix('generated.json')
const GENERATED_FILES = inFix('__generated__')
const FAKE_PACKAGE_JSON = inFix(`fake-pkg.json`)

goodrun(
  ['-i', FAKE_PACKAGE_JSON, '-a', GENERATED, '-o', GENERATED_FILES],
  (done, raw) => {
    expect(stripAnsi(raw)).toMatchSnapshot()
    fork(done)(z => {
      // expect(z).toBeTruthy()
      expect(z).toMatchSnapshot()
      done()
    })(map(JSON.parse, readFile(GENERATED)))
  }
)
afterAll(done => {
  pipe(
    fork(done)(() => {
      done()
    })
  )(both(removeFile(GENERATED))(rimraf(GENERATED_FILES)))
})
