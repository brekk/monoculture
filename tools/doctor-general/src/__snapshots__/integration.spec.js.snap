// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doctor-general -i fixture/fake-pkg.json -a fixture/generated.json -o fixture/__generated__ 1`] = `""`;

exports[`doctor-general -i fixture/fake-pkg.json -a fixture/generated.json -o fixture/__generated__ 2`] = `
[
  {
    "comments": [
      {
        "addTo": "",
        "end": 11,
        "fileGroup": "",
        "filename": "fixture/pkg/one/main.ts",
        "keywords": [
          "@example",
          "@name",
        ],
        "lines": [
          "A curried binary addition function",
          "@name add",
          "@example",
          "\`\`\`ts live=true",
          "import { add } from './fixture'",
          "*",
          "add(10)(20) === 30",
          "\`\`\`",
        ],
        "links": [],
        "start": 2,
        "structure": {
          "example": "\`\`\`ts live=true
import { add } from './fixture'

add(10)(20) === 30
\`\`\`",
          "name": "add",
        },
        "summary": "A curried binary addition function",
      },
    ],
    "filename": "fixture/pkg/one/main.ts",
    "links": [],
    "order": 0,
    "pageSummary": "",
    "slugName": "main",
    "workspace": "pkg",
  },
  {
    "comments": [
      {
        "addTo": "",
        "end": 12,
        "fileGroup": "",
        "filename": "fixture/pkg/three/three.js",
        "keywords": [
          "@example",
          "@link",
          "@name",
          "@see",
        ],
        "lines": [
          "A curried binary addition function",
          "@name add",
          "@see {@link cool}",
          "@example",
          "\`\`\`ts live=true",
          "import { add } from './fixture'",
          "*",
          "add(10)(20) === 30",
          "\`\`\`",
        ],
        "links": [
          "cool",
        ],
        "start": 2,
        "structure": {
          "example": "\`\`\`ts live=true
import { add } from './fixture'

add(10)(20) === 30
\`\`\`",
          "name": "add",
          "see": [
            "cool",
          ],
        },
        "summary": "A curried binary addition function",
      },
    ],
    "filename": "fixture/pkg/three/three.js",
    "links": [
      "cool",
    ],
    "order": 0,
    "pageSummary": "",
    "slugName": "three",
    "workspace": "pkg",
  },
  {
    "comments": [
      {
        "addTo": "",
        "end": 4,
        "fileGroup": "",
        "filename": "fixture/pkg/two/two.ts",
        "keywords": [
          "@page",
        ],
        "lines": [
          "@page minus",
        ],
        "links": [],
        "start": 2,
        "structure": {
          "detail": 2,
          "name": "minus",
          "page": "minus",
        },
        "summary": "@page minus",
      },
      {
        "addTo": "",
        "end": 15,
        "fileGroup": "",
        "filename": "fixture/pkg/two/two.ts",
        "keywords": [
          "@example",
          "@link",
          "@name",
        ],
        "lines": [
          "A curried binary addition function. {@link inlineDetail} Hey there!",
          "@name add",
          "@example",
          "\`\`\`ts live=true",
          "import { add } from './fixture'",
          "*",
          "add(10)(20) === 30",
          "\`\`\`",
        ],
        "links": [
          "inlineDetail",
        ],
        "start": 6,
        "structure": {
          "A": [
            "curried",
            "binary",
            "addition",
            "function.",
            "{link",
            "inlineDetail}",
            "Hey",
            "there!",
          ],
          "example": "\`\`\`ts live=true
import { add } from './fixture'

add(10)(20) === 30
\`\`\`",
          "name": "add",
        },
        "summary": "A curried binary addition function. {@link inlineDetail} Hey there!",
      },
    ],
    "filename": "fixture/pkg/two/two.ts",
    "links": [
      "inlineDetail",
    ],
    "order": 0,
    "pageSummary": "",
    "pageTitle": "minus",
    "slugName": "two",
    "workspace": "pkg",
  },
]
`;
